core/cachemap.go

```diff
@@ -0,0 +1,77 @@
+package core
+
+import (
+	"sync"
+)
+
+type cacheMap[K comparable, T any] struct {
+	cache map[K]T
+	l     sync.Mutex
+}
+
+func newCacheMap[K comparable, T any]() *cacheMap[K, T] {
+	return &cacheMap[K, T]{
+		cache: make(map[K]T),
+	}
+}
+
+func (cache *cacheMap[K, T]) Get(key K) (T, bool) {
+	cache.l.Lock()
+	defer cache.l.Unlock()
+
+	val, found := cache.cache[key]
+	return val, found
+}
+
+type Initializer[T any] interface {
+	// Put sets the value and releases the lock on the cache.
+	Put(val T)
+
+	// Release will release the lock on the cache.
+	Release()
+}
+
+// TODO per-key locks
```

Comment:
In case you haven't seen it before, singleflight does what we need here (iiuc) https://pkg.go.dev/golang.org/x/sync/singleflight

Reply:
Ended up just adjusting my implementation for this after seeing that it's pretty straightforward.